addOns/network/src/main/java/org/zaproxy/addon/network/internal/client/BaseHttpSender.java

```diff
@@ -381,6 +391,44 @@ private boolean isAuthenticationRequired(T1 ctx, HttpMessage message, User user)
                 && !user.isAuthenticated(message);
     }
 
+    private void sendRateLimited(
+            T1 ctx,
+            T2 requestCtx,
+            HttpRequestConfig requestConfig,
+            HttpMessage message,
+            ResponseBodyConsumer<T3> responseBodyConsumer)
+            throws IOException {
+        Pair<RateLimitTrackerEntry, RateLimitParamRule> entryAndRule =
+                rateLimitTracker.getOrCreate(message, ctx.getInitiator());
+        if (entryAndRule == null) {
+            LOGGER.debug("Rate limit not requested for {}", message.getRequestHeader().getURI());
+            sendImpl(ctx, requestCtx, requestConfig, message, responseBodyConsumer);
+            return;
+        }
+        RateLimitTrackerEntry trackerEntry = entryAndRule.first;
+        RateLimitParamRule rule = entryAndRule.second;
+        synchronized (trackerEntry) {
+            long millisToWait =
+                    (1000 / rule.getRequestsPerSecond())
+                            - (System.currentTimeMillis() - trackerEntry.getLastRequestTime());
+            if (millisToWait > 0) {
+                try {
+                    LOGGER.debug(
+                            "{}: sleeping for {} ms",
+                            message.getRequestHeader().getURI(),
+                            millisToWait);
+                    Thread.sleep(millisToWait);
+                } catch (InterruptedException e) {
+                    LOGGER.warn("Request sleep interrupted", e);
+                }
+            }
+            trackerEntry.recordRequest();
+        }
+        rateLimitTracker.fireObserver();
+
+        sendImpl(ctx, requestCtx, requestConfig, message, responseBodyConsumer);
```

Comment:
This logic should be in the `RateLimitTrackerImpl`, the sender should just need to call a method and do nothing else, e.g.:
```suggestion
        rateLimitTracker.rateLimit(message, ctx.getInitiator());
        sendImpl(ctx, requestCtx, requestConfig, message, responseBodyConsumer);
```

Reply:
Not that important since these are internal classes but `RateLimitTracker` could be renamed to `RateLimiter` and the new method (if added) to `limit` or `throttle`.

Reply:
This should be tested.